mantidproject/mantid
# Engin-X full calibration improvements

Fixes #15317

**To test**:
- Check code changes.
- Unit, system and doc tests have been updated and extended.
- Adds new output to EnggCalibrateFull that exposes the parameters fitted for every peak for every detector.
- The documentation of EnggFitPeaks now provides more details and links to functions and algorithms.
- Basic tests of the Engg GUI should go well. For example a basic focus and a basic calibration should still run normally (final calibration results might change slightly).
- The calibration is now more reliable in general. Differences are not big except for 15-30 detectors which were suffering from wrong peak fits. This is not obvious to check (some hints given below).

Some of the changes made here go a bit in the direction of #14596: the long vanadium run used now in the full calibration system test is 194547, to match the ceria run 193749.

To see a full calibration in action (warning may take tens of minutes on slow machines). Note this is not strictly required, as it is tested in a system test:

```
Load('ENGINX00193749.nxs', OutputWorkspace='ws_ceria')
Load('ENGINX00194547.nxs', OutputWorkspace='ws_vana')

posTable_all, peaksTable_all = EnggCalibrateFull(Workspace='ws_ceria', VanadiumWorkspace='ws_vana', SpectrumNumbers='1-2400', OutDetPosFilename='enginx_calibrate_full_ceria193749_both_banks.csv')
```

And to inspect the results, for example:

```
tbl=mtd['posTable_all']
L2=tbl.column('\delta L2 (calibrated - old)')
# check differences in L2
min(L2), sum(L2)/len(L2), max(L2)
# find the max difference for example
val, idx = min((val, idx) for (idx, val) in enumerate(L2))
val, idx

# Check position and parameters for one detector example (spectrum 802):
postTable_one, peaksTable_one  = EnggCalibrateFull(Workspace='ws_ceria', VanadiumWorkspace='ws_vana', SpectrumNumbers='802', OutDetPosFilename='enginx_calibrate_full_ceria193749_one.csv')

# To see statistics about the detectors (using the .csv file that EnggCalibrateFull produces):
import numpy as np
import matplotlib.pyplot as plt
calib = genfromtxt('enginx_calibrate_full_ceria193749_both_banks.csv', delimiter=',')
calib.shape

L2_delta = calib[:,10]
difc = calib[:,11]
tzero = calib[:, 12]
# check percentiles:
np.percentile(tzero, 99)
# cumulative distribution plots
def cdf_plot(data):
    values, base = np.histogram(data, bins=100)
    cumulative = np.cumsum(values)
    plt.plot(base[:-1], cumulative, c='blue')
    plt.show()

cdf_plot(difc)
cdf_plot(tzero)
cdf_plot(L2_delta)
```
